jnettop.8
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH JNETTOP 8 "April 8, 2006"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
jnettop \- View hosts/ports taking up the most network traffic
.SH SYNOPSIS
.B jnettop
[options] [-i interface] [-d filename] [-f filename] [-x rule]
.SH DESCRIPTION
This manual page documents briefly the
.B jnettop
command. This manual page is OBSOLETE. Please use jnettop -h as a main source of information about usage.
.PP
.\" TeX users may be more comfortable with the \fB<whatever>\fP and
.\" \fI<whatever>\fP escape sequences to invode bold face and italics, 
.\" respectively.
\fBjnettop\fP captures traffic coming across the host it is running on and displays streams sorted by bandwidth they use. Result is a nice listing of communication on network by host and port, how many bytes went through this transport and the bandwidth it is consuming.
.SH OPTIONS
These programs follow the usual GNU command line syntax, with long
options starting with two dashes (`-').
A summary of options is included below.
.TP
.B \-h, \-\-help
Show summary of options.
.TP
.B \-v, \-\-version
Show version of program.
.TP
.B \-b, \-\-bit-units
show BPS In bits per second, not bytes per second
.TP
.B \-c, \-\-content-filter
disable content filtering
.TP
.B \-d, \-\-debug filename
write debug information into file
.TP
.B \-f, \-\-config-file filename
reads configuration from filename. defaults to ~/.jnettop. an example can be found at /usr/share/doc/jnettop/dot.jnettop.
or in .jnettop file from original distribution package.
.TP
.B \-i, \-\-interface name
capture packets on specified interface
.TP
.B \-\-local-aggr [none|host|port]
set local aggregation to specified value
.TP
.B \-n, \-\-no-resolver
disable resolving of ip addresses
.TP
.B \-p, \-\-promiscuous
enables promiscuous mode on the sniffed interface
.TP
.B \-\-remote-aggr [none|host|port]
set remote aggregation to specified value
.TP
.B \-s, \-\-select-rule name
selects one of the rules defined in .jnettop configuration file (by it's name)
.TP
.B \-x, \-\-filter rule
allows for specification of custom filtering rule. this allows for
.BR tcpdump(1)
style syntax. don't forget to enclolse the filter into quotes when running from a shell.
.SH CONFIGURATION
Program looks for settings in the file specified by parameter \fB-f\fP, which defaults to ~/.jnettop. Configuration file is
an ordinary text file with keywords and their arguments. You HAVE to enclose arguments into double quotes. Available keywords
are:
.TP
\fBinterface "<interface_name>"\fR
The \fBinterface\fP keyword specifies network interface on which to start listening. Example:
.RS
.PP
interface "eth0"
.RE
.TP
\fBlocal_aggregation [none|host|port]\fR
The \fBlocal_aggregation\fP keyword specifies initial active local aggregation. Valid values are \fBnone\fP, \fBhost\fP and \fBport\fP. Example:
.RS
.PP
local_aggregation host
.RE
.TP
\fBpromisc [on|off]\fR
The \fBpromisc\fP keyword specifies, whether jnettop captures packets in promiscuous mode. Example:
.RS
.PP
promisc on
.RE
.TP
\fBremote_aggregation [none|host|port]\fR
The \fBremote_aggregation\fP keyword specifies initial active remote aggregation. Valid values are \fBnone\fP, \fBhost\fP and \fBport\fP. Example:
.RS
.PP
remote_aggregation port
.RE
.TP
\fBresolve [on|off]\fR
The \fBresolve\fP keyword specifies, whether resolving is performed on the IP addresses or not.
.RS
.PP
resolve off
.RE
.TP
\fBresolve_rule "<network address>" "<network mask>" [normal|external] (<arguments> ...)\fR
The \fBresolve_rule\fP keyword adds one resolver into list of resolvers for specified address. When resolving, jnettop examines all the rules in the
order how they were specified in configuration file. If the network address matches specified range, declared resolver is used. Resolver can be
\fBnormal\fP, which means the standard DNS lookup or \fBexternal\fP, which executes specified external program to perform resolving. This can
be used with bundled jnettop-lookup-nmb script, which looks up IP addresses using nmblookup(1) tool. If a tool returns empty string or DNS
is not found, next rule is examined. If jnettop runs out of rules, than the standard DNS lookup is executed.
.RS
.PP
resolve_rule "192.168.0.0" "255.255.255.0" normal
.br
resolve_rule "192.168.0.0" "255.255.255.0" external "/usr/share/jnettop/jnettop-lookup-nbm"
.RE
.TP
\fBrule "<rule_name>" "<rule_definition>"\fR
The \fBrule\fP keyword defines a set of predefined tcpdump(1)-like filters to apply. You can specify various filters as "show me what 192.168.1.32" sends:
.RS
.PP
rule "show 192.168.1.32" "src 192.168.1.32"
.RE
.TP
\fBselect_rule "<rule_name>"\fR
The \fBselect_rule\fP keyword specifies initial active predefined rule. The rule must be defined before this keyword is used. Example:
.RS
.PP
select_rule "show 192.168.1.32"
.RE
.TP
\fBvariable "<variable_name>" "<variable_contents>"\fR
The \fBvariable\fP keyword introduces a string variable for use in future rule definitions. It can be used to shorten rule definitions. Example:
.RS
.PP
variable "intranet" "net 192.168.0.0/16 or 10.0.0.0/8 or 172.16.0.0/12"
.RE
.TP
For more information, see README file or .jnettop example configuration file included in distribution.
.SH AUTHOR
This manual page was originally written by Ari Pollak <ari@debian.org>,
for the Debian GNU/Linux system. Small changes were
introduced by Jakub Skopal <j@kubs.cz>



NEWS
2006-04-11:	Added support for UIA output
		Added support for conditional compilation of NCURSES and UIA
		Added optional support for syslog
		Aded support for explicit list of local networks

2006-03-31:	Added text output
		Added support for IEEE 8021.g VLAN

2005-07-01:	Added pause
		Added sorting
		Added external resolvers
		Added -n command line parameter (disable resolving)

2005-06-30:	Refactoring to more reusable design

2004-10-04:	Added --disable-multithreaded-resolver option for FreeBSD
		Fixed more porting issues.

2004-10-02:	Fixed porting issues to freebsd and solaris

2004-09-28:	Added packet counting instead of just bytes counting
		Added support for raw encapsulation present on linux IPv6 tunnel devices

2004-08-27:	Added IPv6 support
		Address field width is now dynamic

2003-07-30:	Added "-p" command line parameter (promiscuous mode)
		Added "local_aggregation", "remote_aggregation" and "select_rule" keywords to config file (see README)
		Added "--local-aggr", "--remote-aggr" and "-s" command line parameters

2003-04-23:	Added man page

2003-04-16:	Added "interface" keyword
		
2002-10-16:	Added suspend feature

2002-10-13:	Added -f command line parameter
		Added support for BPF filters to filter packets prior to processing them

2002-08-31:	I added -b switch for bits/bytes per second

2002-08-28:	Package is now compilable under FreeBSD, Solaris, Debian and on
		various CPUs. See PORTING file.

2002-08-27:	Added -i command line parameter
		Added -d command line parameter
		jnettop now correctly interprets traffic on linux "any" interface
		separate in/out accounting

2002-08-22:	The package is entirely new.

$Header: /home/jakubs/DEV/jnettop-conversion/jnettop/NEWS,v 1.16 2006-04-11 15:21:05 merunka Exp $


jnettop.c
/*
 *    jnettop, network online traffic visualiser
 *    Copyright (C) 2002-2006 Jakub Skopal
 *
 *    This program is free software; you can redistribute it and/or modify
 *    it under the terms of the GNU General Public License as published by
 *    the Free Software Foundation; either version 2 of the License, or
 *    (at your option) any later version.
 *
 *    This program is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License
 *    along with this program; if not, write to the Free Software
 *    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 *    $Header: /home/jakubs/DEV/jnettop-conversion/jnettop/jnettop.c,v 1.38 2006-04-12 07:47:01 merunka Exp $
 *
 */

#include "jbase.h"
#include "jdevice.h"
#include "jcapture.h"
#include "jprocessor.h"
#include "jresolver.h"
#include "jresolv.h"
#include "jfilter.h"
#include "jutil.h"
#include "jconfig.h"
#include "jcursesdisplay.h"
#include "jtxtdisplay.h"
#include "juiadisplay.h"

#define			DEBUGOUT_NONE	0
#define			DEBUGOUT_SYSLOG	1
#define			DEBUGOUT_FILE	2
int			debugOut = DEBUGOUT_NONE;
FILE *			debugFile = NULL;

volatile int		threadCount;

jbase_display *		currentDisplay;

void debug(int priority, const char *format, ...) {
	static char buffer[32768];
	va_list ap;
	va_start(ap, format);
	vsprintf(buffer, format, ap);
	va_end(ap);

	switch (debugOut) {
		case DEBUGOUT_FILE:
			fprintf(debugFile, "%d - %d, %s\n", getpid(), priority, buffer);
			break;
#ifdef SUPPORT_SYSLOG
		case DEBUGOUT_SYSLOG:
			syslog(priority, "%d - %d, %s\n", getpid(), priority, buffer);
#endif
	}
}

void jbase_cb_DrawStatus(const gchar *msg) {
	currentDisplay->drawstatus(msg);
}

void parseCommandLineAndConfig(int argc, char ** argv) {
	char * configFileName = NULL;
	char * selectRuleName = NULL;
	int a;

	jconfig_Setup();
	
	for (a=1; a<argc; a++) {
		if (!strcmp(argv[a], "-v") || !strcmp(argv[a], "--version")) {
			printf(PACKAGE_STRING "\nWritten by Jakub Skopal <j@kubs.cz>\n\nSee copyright in the COPYING file.\n");
			exit(0);
		}
		if (!strcmp(argv[a], "-h") || !strcmp(argv[a], "--help")) {
			printf(	"Usage: jnettop [-hv] [-i interface] [-d filename]\n"
				"\n"
				"    -h, --help             display this help message\n"
				"    -v, --version          display version information\n\n"
				"    -b, --bit-units        show BPS in bits per second, not bytes per second\n"
				"    -c, --content-filter   disable content filtering\n"
				"    -d, --debug filename   write debug information into file (or syslog)\n"
				"    --display type         type of display (curses, text, uia)\n"
				"    -f, --config-file name reads configuration from file. defaults to ~/.jnettop\n"
				"    --format format        list of fields to list in text output\n"
				"    -i, --interface name   capture packets on specified interface\n"
				"    --local-aggr arg       set local aggregation to none/host/port\n"
				"    -n, --no-resolver      disable resolving of addresses\n"
				"    -p, --promiscuous      enable promisc mode on the devices\n"
				"    --remote-aggr arg      set remote aggregation to none/host/port\n"
				"    -s, --select-rule rule selects one of the rules defined in config file\n"
				"                           by it's name\n"
				"    -t, --timeout sec      timeout in seconds after which jnettop ends (text display)\n"
				"    -x, --filter rule      allows for specification of custom filtering rule\n"
				"                           this follows tcpdump(1) syntax. don't forget to\n"
				"                           enclose the filter into quotes when running from shell\n"
				"\n"
				"Report bugs to <j@kubs.cz>\n"
				"\n"
				"    Format variable can be CSV (comma separated values), TSV (tab separated values)\n"
				"    or completelly custom format string, where the following identifiers are subst-\n"
				"    ituted when surrounded by '$':\n"
				"       src, srcname, srcport, srcbytes, srcpackets, srcbps, srcpps,\n"
				"       dst, dstname, dstport, dstbytes, dstpackets, dstbps, dstpps,\n"
				"       proto, totalbytes, totalpackets, totalbps, totalpps, filterdata\n"
				"\n"
				"    example:\n"
				"       jnettop --display text -t 5 --format CSV\n"
				"       jnettop --display text -t 5 --format '$srcname$,$srcport$,$dstname$,$dstport$,$totalbps$'\n"
				"\n"
			);
			exit(0);
		}
		if (!strcmp(argv[a], "-c") || !strcmp(argv[a], "--content-filter")) {
			jconfig_Settings.onoffContentFiltering = FALSE;
			continue;
		}
		if (!strcmp(argv[a], "--display")) {
			if (a+1>=argc) {
				fprintf(stderr, "%s switch requires argument\n", argv[a]);
				exit(255);
			}
			++a;
			if (jcursesdisplay_Functions.supported && !strcmp(argv[a], "curses")) {
				currentDisplay = &jcursesdisplay_Functions;
			} else if (jtxtdisplay_Functions.supported && !strcmp(argv[a], "text")) {
				currentDisplay = &jtxtdisplay_Functions;
			} else if (juiadisplay_Functions.supported && !strcmp(argv[a], "uia")) {
				currentDisplay = &juiadisplay_Functions;
			} else {
				fprintf(stderr, "display type %s is not supported.\n", argv[a]);
				exit(255);
			}
			continue;
		}
		if (!strcmp(argv[a], "-i") || !strcmp(argv[a], "--interface")) {
			if (a+1>=argc) {
				fprintf(stderr, "%s switch requires argument\n", argv[a]);
				exit(255);
			}
			jconfig_Settings.deviceName = argv[++a];
			continue;
		}
		if (!strcmp(argv[a], "-s") || !strcmp(argv[a], "--select-rule")) {
			if (a+1>=argc) {
				fprintf(stderr, "%s switch requires argument\n", argv[a]);
				exit(255);
			}
			selectRuleName = argv[++a];
			continue;
		}
		if (!strcmp(argv[a], "-d") || !strcmp(argv[a], "--debug")) {
			if (a+1>=argc) {
				fprintf(stderr, "%s switch requires filename to debug to as an argument\n", argv[a]);
				exit(255);
			}
			++a;
			if (!strcmp(argv[a], "syslog")) {
#ifdef SUPPORT_SYSLOG
				debugOut = DEBUGOUT_SYSLOG;
#else
				fprintf(stderr, "Syslog output not enabled in compilation\n");
				exit(255);
#endif
			} else {
				debugFile = fopen(argv[a], "w");
				if (!debugFile) {
					perror("Could not open debug file");
					exit(255);
				}
				debugOut = DEBUGOUT_FILE;
			}
			continue;
		}
		if (!strcmp(argv[a], "-f") || !strcmp(argv[a], "--config-file")) {
			if (a+1>=argc) {
				fprintf(stderr, "%s switch required argument\n", argv[a]);
				exit(255);
			}
			configFileName = argv[++a];
			continue;
		}
		if (!strcmp(argv[a], "-x") || !strcmp(argv[a], "--filter")) {
			const char *ret;
			char *commandLineRule;
			if (a+1>=argc) {
				fprintf(stderr, "%s switch requires argument\n", argv[a]);
				exit(255);
			}
			commandLineRule = argv[++a];
			ret = jutil_ValidateBPFFilter(commandLineRule);
			if (ret) {
				fprintf(stderr, "Error compiling rule: %s\n", ret);
				exit(255);
			}
			JCONFIG_BPFFILTERS_SETSELECTEDFILTER(JCONFIG_BPFFILTERS_LEN);
			jconfig_AddBpfFilter("<commandline>", commandLineRule);
			continue;
		}
		if (!strcmp(argv[a], "-p") || !strcmp(argv[a], "--promiscuous")) {
			jconfig_Settings.onoffPromisc = TRUE;
			continue;
		}
		if (!strcmp(argv[a], "-n") || !strcmp(argv[a], "--no-resolve")) {
			jconfig_Settings.onoffResolver = FALSE;
			continue;
		}
		if (!strcmp(argv[a], "--local-aggr")) {
			if (a+1>=argc || (jconfig_Settings.localAggregation = jutil_ParseAggregation(argv[++a]))==-1) {
				fprintf(stderr, "%s switch requires none, host or port as an argument\n", argv[a]);
				exit(255);
			}
			continue;
		}
		if (!strcmp(argv[a], "--remote-aggr")) {
			if (a+1>=argc || (jconfig_Settings.remoteAggregation = jutil_ParseAggregation(argv[++a]))==-1) {
				fprintf(stderr, "%s switch requires none, host or port as an argument\n", argv[a]);
				exit(255);
			}
			continue;
		}
		{
			int consumed = currentDisplay->processargument((const gchar **) argv+a, argc-a);
			if (consumed) {
				a += consumed - 1;
				continue;
			}
		}
		fprintf(stderr, "Unknown argument: %s\n", argv[a]);
		exit(255);
	}

	if (!jconfig_ParseFile(configFileName)) {
		exit(255);
	}

	jconfig_SetDefaults();

	if (selectRuleName) {
		int i = jconfig_FindBpfFilterByName(selectRuleName);
		if (i == -1) {
			fprintf(stderr, "Rule '%s' specified on the command line is not defined.\n", selectRuleName);
			exit(255);
		}
		JCONFIG_BPFFILTERS_SETSELECTEDFILTER(i);
	}
}

void initializeDevices() {
	if (!jdevice_LookupDevices()) {
		exit(255);
	}
	
	if (!jdevice_DevicesCount) {
			if (!jconfig_Settings.deviceName) {
				fprintf(stderr, "Autodiscovery found no devices. Specify device you want to watch with -i parameter\n");
				exit(255);
			}
			if (!(jconfig_Settings.device = jdevice_CreateSingleDevice(jconfig_Settings.deviceName))) {
				exit(255);
			}
	} else if (jconfig_Settings.deviceName) {
		int i;
		for (i=0; i<jdevice_DevicesCount; i++) {
			if (!strcmp(jdevice_Devices[i].name, jconfig_Settings.deviceName)) {
				jconfig_Settings.device = jdevice_Devices + i;
				break;
			}
		}

		if (i >= jdevice_DevicesCount) {
			if (!(jconfig_Settings.device = jdevice_CreateSingleDevice(jconfig_Settings.deviceName))) {
				exit(255);
			}
		}
	}

	if (!jconfig_Settings.device) {
		jconfig_Settings.deviceName = jdevice_Devices[0].name;
		jconfig_Settings.device = jdevice_Devices;
	}

	if (!jdevice_CheckDevices()) {
		exit(255);
	}
}

int main(int argc, char ** argv) {
	g_thread_init(NULL);

	jcapture_Setup();
	jprocessor_Setup();
	jresolver_Setup();

	if (jcursesdisplay_Functions.supported)
		currentDisplay = &jcursesdisplay_Functions;
	else
		currentDisplay = &jtxtdisplay_Functions;

	parseCommandLineAndConfig(argc, argv);

	if (!currentDisplay->presetup()) {
		return 0;
	}

	jconfig_ConfigureModules();
	initializeDevices();

	currentDisplay->setup();

	while (TRUE) {

		jprocessor_ResetStats();

		jcapture_SetDevice(jconfig_Settings.device);
		jcapture_SetBpfFilterText(jconfig_GetSelectedBpfFilterText());

		currentDisplay->prerun();

		jcapture_Start();
		jprocessor_Start();

		if (!currentDisplay->run()) {
			// In case we're not switching to another device, we can happily finish
			// after our display thread dies. (mind the endwin())
			break;
		}

		jcapture_Kill();
		
		while (threadCount) {
			g_thread_yield();
		}
	}

	if (debugFile) {
		fclose(debugFile);
	}

	currentDisplay->shutdown();
	return 0;
}
